### ISSUE TITLE
Framework or tool available to run LTP tests

### ISSUE THREAD

**Issue by chetjain** [linux-test-project/ltp#663]
General Query:
---------------
Under "Integrated Software & System Test" Environment, we need to be running multiple LTP testcases concurrently. This would induce kernel stress from multiple OS parameters.
This might include running mix of BASE(Kernel|IPC|Threads,Syscall etc), IO, TCP & NFS tests.

Do we have some known mechanism, how to achieve this? (not runltp -x)
Currently, we have achieved this within IBM test environment through in-house tool and would like to see, if we can open source and invite other collaborators to use it. (Please recommend, if you have some git project for us to contribute)
This methodology has introduced several crashes and I'm sure, this would benefit all Linux Distributor partners.

Running LTP through avocado/autotest/any framework would be beneficial for functional validation only. It might trigger the tests in parallel, but it will not monitor & control the execution run and also provide consolidated execution summary.

Something like this.
***************************************
PASS%=83
SKIP%=6
TBROK%=2
TCONF%=3
***************************************
- - - - - - - - - - -  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
TEST_CASE | TOTAL_ITR(7950) | TOTAL_PASS(6159) | TOTAL_FAIL(1252) | TOTAL_SKIP(539) | TOTAL_TBROK(175) | TOTAL_CONF(317) | PASS_RATE(83)
- - - - - - - - - - -  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
waitpid12|30|30|0|0|0|0|100
sendfile03_64|81|81|0|0|0|0|100
signalfd4_02|76|76|0|0|0|0|100
open10|122|122|0|0|0|0|100

********************************************
SAMPLE LIST OF LTP SCENARIO's
********************************************
[2020/03/05,18:41:39] [start.ltp.dirty.bk] [notice] Scenario_1:  flock02(96) sched_setparam05(53) setregid01(44) mbind04(106) statx04(1) dirtyc0w(120) mountns02(124) recvmsg03(58) llistxattr03(100) 
[2020/03/05,18:43:57] [start.ltp.dirty.bk] [notice] Scenario_2:  fcntl33_64(70) userns07(114) pidns02(126) mknod09(100) getrlimit03(81) profil01(81) kill11(8) 
[2020/03/05,18:44:59] [start.ltp.dirty.bk] [notice] Scenario_3:  fanotify12(29) mount03(63) 
[2020/03/05,18:48:08] [start.ltp.dirty.bk] [notice] Scenario_4:  fcntl11_64(21) getsid01(116) recv01(46) execve01(50) statx01(53) ioprio_set03(124) fs_inod(1) dup02(9) crash01(1)

Appreciate any help on this. Thanks

**Comment by metan-ucw** [comment]
See:
https://people.kernel.org/metan/towards-parallel-kernel-test-runs

**Comment by chetjain** [comment]
Hey Thanks for responding quickly.
We tried runltp-ng yesterday to use it on local host. We found, the following limitation:

(a)Testcases under testsuite are not running in parallel, when specified using --cmd. It is again a serial   execution  
(b) --cmd cannot take multiple arguments
(c) Summary given only from 1 execution. When ran multiple times, no consolidation happening under same LOG file
EX: move_pages01 can be run in combination with different testcases multiple times. It might fail/pass, when attempted under different circumstances
(d) There is no method available to run mix of testcases from 2 or more different testsuites. 

Generally we prefer to run the tests for upto 3 days and during that time, each test case might have run several times under  RANDOM intervals of time. We can't track that level of execution here,

We see, this is useful for triggering the tests on multiple REMOTE host, through centralized server.
Please suggest.

**Comment by metan-ucw** [comment]
(a) Testcases cannot be run in parallel because the constraints are not yet propagated from the tests to the execution framework, this is an ongoing effort that will take at least a few more months to finish. Unless we know which tests are safe to run in parallel there is no point in running tests in parallel.

(b) That's one would be easy to fix.

(c) What do you mean by consolidation? The number of pass/faill/... is collected under the test name so if you run a testcase 10 times the counters should show that.

(d) Should be also reasonably easy to fix.

**Comment by chetjain** [comment]
(a) This would be the most challenging exercise for anyone and need multiple run-time effort to understand the behavior. I have been working on this from last 1 year to create POC for our stress run. 
Currently, our methodology follows certain rules. Few are like,
(1) Avoid running more than 1 resource intensive tests at a time
RIT="*stress*|aio*|*mem*|*fork*|*mmap*|*cpu*|*oom*|*cgroup*"
(2) Avoid running Memory exploiting tests
(3) Remove network tests that creates loop-back ltp interfaces and cause lab storm
(4) Avoid TC, that require multiple interface etc.,
(5) Do not start more that 4 concurrent NW/NFS tests at a time
(6) Define the number of ITERATIONs based on test selection
(7) Define the number of concurrent tests based on available CPU/Mem/IO etc.,

Appreciate if you could share your methodology and thinking. This would help us to improve our tests, But, with our current method, there are numerous bugs been filed. We had interlock with SUSE architects and share the list under NDA. They are now exploring to use this in their ppc test env.

(c) Currently, the iteration/time  (-I & -t) attribute is not yet inherited to runltp-ng. So, I could not able to find a method to run the test more than once and verify. 
However, there is a possibility that, this TC would be called at different interval of time. 
For ex., There are 3 Scenario's listed down here ran at 3 different time intervals.
 mknod07 ran 37 times in scenario_9, 38 time in scenario_258 and 1 time during scenario517.

************************************************************************************
[2020/03/05,15:22:07] [start.ltp.dirty.bk] [notice] **Scenario_9**: **mknod07(37)** sendmmsg01(99) capset03(122)
[2020/03/07,15:09:18] [start.ltp.dirty.bk] [notice] **Scenario_258**:  vma03(22) removexattr02(35) 
set_thread_area01(63) **mknod07(38)** vmsplice01(106) setsockopt02(125) 
[2020/03/09,06:06:08] [start.ltp.dirty.bk] [notice] **Scenario_517**:  getgid01(124) setsockopt01(22) **mknod07(1)** ar01.sh(20) posix_fadvise03_64(47) 
************************************************************************************
If we need a consolidated result, then it would be 
TEST_CASE | TOTAL_ITR| TOTAL_PASS  | TOTAL_FAIL  | TOTAL_SKIP  | TOTAL_TBROK  | TOTAL_CONF | PASS_RATE 
mknod07|76|76|0|0|0|0|100

(e) Apart from all of this, we also need to repeat same pattern/order of tests during problem recreation. So, we might need an option of providing scenario file as an input, (like the format, showed in (c) above)

**Comment by metan-ucw** [comment]
(a) At this moment I'm working on propagating the test requirements from the testcases, I think the blog post I posted here and the links there explain in great detail what I'm trying to achieve. In short I managed to write a code that generates a json file that contains test requirements parsed from the test source code when LTP is being build. Next step is making use of that information. For example if LTP test has '.restore_wallclock' flag set in the test structure it means that it modifies system clock and that tests with this flag cannot be run in parallel. So in a future I'm aiming for fine grade ruleset that would allow us to run as much test in parallel as possible. I also had a [presentation](https://elinux.org/File:Test_metadata_extraction.pdf) on the test automation summit. For a start I'm trying to get the syscall tests running in parallel but I do not think that it will happen soon. I will be happy if I'm going to get the metadata generator merged upstream in a few months, then I will try to make use of the metadata in the runltp-ng.
